examples/javascript/groveeldriver.js

```diff
@@ -0,0 +1,54 @@
+/*jslint node:true, vars:true, bitwise:true, unparam:true */
+/*jshint unused:true */
+/*
+* Author: Zion Orent <zorent@ics.com>
+* Copyright (c) 2015 Intel Corporation.
+*
+* Permission is hereby granted, free of charge, to any person obtaining
+* a copy of this software and associated documentation files (the
+* "Software"), to deal in the Software without restriction, including
+* without limitation the rights to use, copy, modify, merge, publish,
+* distribute, sublicense, and/or sell copies of the Software, and to
+* permit persons to whom the Software is furnished to do so, subject to
+* the following conditions:
+*
+* The above copyright notice and this permission notice shall be
+* included in all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+* NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+* LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+* OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+* WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+var eldriver_lib = require("jsupm_groveeldriver");
+
+// The was tested with the Grove El Driver Module
+// Instantiate a Grove El Driver on digital pin D2
+var eldriver_obj = new eldriver_lib.GroveElDriver(2);
+
+var lightState = true;
+
+var myInterval = setInterval(function()
+{
+	if (lightState)
+		eldriver_obj.on();
+	else
+		eldriver_obj.off();
+	lightState = !lightState;
+}, 1000);
+
+// When exiting: turn off EL wire, clear interval, and print message
+process.on('SIGINT', function()
+{
+	clearInterval(myInterval);
+	eldriver_obj.off();
+	eldriver_obj = null;
+	eldriver_lib.cleanUp();
+	eldriver_lib = null;
+	console.log("Exiting...");
+	process.exit(0);
```

Comment:
> In testing upm, we found that GC wasn't called-- neither on scope exit nor on program exit. 

This would be an egregious memory leak, but not one that's the result of running a JavaScript program via node.js. Consider tools like Forever (or anything similar) that automatically restart node.js programs that have terminated unexpectedly: if memory isn't implicitly freed on termination, repeated restarts would cripple the system.


> I wouldn't make a 1-to-1 comparison between JavaScript behavior in a browser and JavaScript behavior in nodeJS. 

Yes, neither would I. I've also been actively developing software for node.js since 2010—so I have a significant amount of experience with the platform. 


> Pretty much all browsers perform correct GC, and browser JavaScript likely hasn't required any type of variable destruction for quite a while.

The "browser" has nothing to do with the JavaScript runtime's GC behaviour, aside from providing the runtime (V8, SpiderMonkey, JCS, Chakra, etc) with metrics that inform its scheduling and execution of a GC cycle. I gave the example of IE8 to illustrate the pointlessness of exposing and invoking GC explicitly. 


> But nodeJS is the wild west regarding GC. As of node 0.10.0, any attempt at guessing when to collect garbage was removed. 

From the [release announcement, almost exactly two years ago](http://blog.nodejs.org/2013/03/11/node-v0-10-0-stable/) that you cited: 

>> One of the toughest things to get right in a garbage collected language is garbage collection. In order to try to avoid excessive memory usage, Node used to try to tell V8 to collect some garbage whenever the event loop was idle.

>> However, knowing exactly when to do this is extremely difficult. There are different degrees of "idleness", and if you get it wrong, you can easily end up spending massive amounts of time collecting garbage when you'd least expect. In practice, disabling the IdleNotification call yields better performance without any excessive memory usage, because V8 is pretty good at knowing when it's the best time to run GC.

>> So, in v0.10, we just ripped that feature out. (According to another point of view, we fixed the bug that it was ever there in the first place.) As a result, latency is much more predictable and stable. You won't see a difference in the benchmarks as a result of this, but you'll probably find that your app's response times are more reliable.

Node.js was trying to be too aggressive, so it stopped doing that and allows V8 to run GC on its own schedule. 

> This leaves us with V8 GC-- which isn't guaranteed.

Indeterminate is not the same as "isn't guaranteed."

> Essentially, v8 was built by Google for Google Chrome. 

It's clear that you're not familiar with [the work](https://github.com/tc39/test262/pulls) I'm [involved in](https://github.com/rwaldron/tc39-notes). 

> Everybody else using v8 has to deal with issues like this one from time to time.

I should've been forthright with you in my first comment: I consulted directly with V8 implementation engineers before posting, to make sure that my generalized understanding of V8 memory management was both current and correct. I wouldn't have posted anything at all if I was wrong. We discussed your reply and that has informed my response, so please accept that I'm responding with correct information from the most credible source.